Support switching burn-wgpu between WebGPU implementations (wgpu <-> Dawn)

p1-0tr opened the pull request:
### Checklist

- [ ] Confirmed that `run-checks all` script has been executed.
- [ ] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Split from: https://github.com/tracel-ai/burn/pull/1475

### Changes

Currently Burn can only be built against the wgpu WebGPU implementation. I found the ability to switch to Dawn useful, e.g. to get float16 support, which helps when trying to run Llama2 on top of Burn.

I've:
* split out any uses of wgpu into a separate module, and defined traits which provide all operations needed by burn-wgpu,
* added Dawn as a submodule,
* modified the build system to be able to build, generate bindings for, and link against Dawn,
* added a module which wraps the usage of the Dawn bindings to make it look more like wgpu.

### Testing

I ran unit tests for the `burn-wgpu` crate. Dawn test cases should be generated when the `dawn` feature is enabled for `burn-wgpu`, some of them will fail because Dawn performs out of bound accesses if no explicit bounds checks are emitted for I/O arrays (I've pushed a branch with a fix for that here - https://github.com/p1-0tr/burn/tree/ps-allow-using-dawn-and-wgpu-w-bounds).

nathanielsimard reviewed (commented):
Very good job, I like how you abstracted the WebGpu Api. I have a few comments, but it's still in draft so some might be already obvious to you.

p1-0tr commented on crates/burn-wgpu/build.rs:
```diff
@@ -0,0 +1,202 @@
+#[cfg(all(feature = "dawn", not(target_os = "macos")))]
+compile_error!("The 'dawn' backend currently only builds on macos.");
```
This was the main reason I left the PR in draft. Linux build should be simple to add, Windows may be more problematic. I'm planning to look into that tomorrow or on Thursday.

p1-0tr commented on crates/burn-wgpu/build.rs:
```diff
@@ -0,0 +1,202 @@
+    println!("cargo:rustc-link-lib=static=tint_utils_cli");
+    println!("cargo:rustc-link-lib=static=tint_utils_command");
+    println!("cargo:rustc-link-lib=static=tint_utils_containers");
+    println!("cargo:rustc-link-lib=static=tint_utils_debug");
+    println!("cargo:rustc-link-lib=static=tint_utils_diagnostic");
+    println!("cargo:rustc-link-lib=static=tint_utils_file");
+    println!("cargo:rustc-link-lib=static=tint_utils_generator");
+    println!("cargo:rustc-link-lib=static=tint_utils_ice");
+    println!("cargo:rustc-link-lib=static=tint_utils_id");
+    println!("cargo:rustc-link-lib=static=tint_utils_macros");
+    println!("cargo:rustc-link-lib=static=tint_utils_math");
+    println!("cargo:rustc-link-lib=static=tint_utils_memory");
+    println!("cargo:rustc-link-lib=static=tint_utils_reflection");
+    println!("cargo:rustc-link-lib=static=tint_utils_result");
+    println!("cargo:rustc-link-lib=static=tint_utils_rtti");
+    println!("cargo:rustc-link-lib=static=tint_utils_socket");
+    println!("cargo:rustc-link-lib=static=tint_utils_strconv");
+    println!("cargo:rustc-link-lib=static=tint_utils_symbol");
+    println!("cargo:rustc-link-lib=static=tint_utils_text");
+    println!("cargo:rustc-link-lib=static=tint_utils_traits");
+    println!("cargo:rustc-link-search={dawn_build_dir}/third_party/abseil/absl/strings");
+    println!("cargo:rustc-link-lib=static=absl_str_format_internal");
+    println!("cargo:rustc-link-lib=static=absl_strings");
+    println!("cargo:rustc-link-lib=static=absl_strings_internal");
+    println!("cargo:rustc-link-search={dawn_build_dir}/third_party/abseil/absl/base");
+    println!("cargo:rustc-link-lib=static=absl_base");
+    println!("cargo:rustc-link-lib=static=absl_spinlock_wait");
+    println!("cargo:rustc-link-lib=static=absl_throw_delegate");
+    println!("cargo:rustc-link-lib=static=absl_raw_logging_internal");
+    println!("cargo:rustc-link-lib=static=absl_log_severity");
+    println!("cargo:rustc-link-search={dawn_build_dir}/third_party/abseil/absl/numeric");
+    println!("cargo:rustc-link-lib=static=absl_int128");
+    println!("cargo:rustc-link-search={dawn_build_dir}/third_party/abseil/absl/hash");
+    println!("cargo:rustc-link-lib=static=absl_city");
+    println!("cargo:rustc-link-lib=static=absl_hash");
+    println!("cargo:rustc-link-lib=static=absl_low_level_hash");
+    println!("cargo:rustc-link-search={dawn_build_dir}/third_party/abseil/absl/container");
+    println!("cargo:rustc-link-lib=static=absl_hashtablez_sampler");
+    println!("cargo:rustc-link-lib=static=absl_raw_hash_set");
+    println!("cargo:rustc-link-lib=dylib=c++");
```
They instruct the linker which libs should be linked against when Dawn is built. Some of them are probably redundant (I added all the `*.a`s that Dawn generates), but I'm going to work down the list to objects that are actually required.

p1-0tr commented on crates/burn-wgpu/src/compute/server.rs:
```diff
@@ -197,64 +184,30 @@ where
-        }
-
-        #[cfg(not(target_family = "wasm"))]
         Reader::Concrete(self.buffer_reader(handle).read(&self.device))
```
Ah, no, I must have missed this when resolving a rebase. Thanks :)

p1-0tr replied:
BTW, is there a way to run unit tests when targetting wasm?

p1-0tr replied:
Should, be fixed now :)

nathanielsimard replied:
I don't think we run tests yet on wasm, but it should be possible

dcvz commented on crates/burn-wgpu/src/compiler/wgsl/base.rs:
```diff
@@ -218,7 +218,9 @@ impl Display for Variable {
-                Elem::I32 => f.write_fmt(format_args!("{number}i")),
+                // Dawn seems to get tripped up by the 'i' suffix, while wgpu is happy
+                // with or without it, so emit the literal without it.
+                Elem::I32 => f.write_fmt(format_args!("{number}")),
```
sanity question: wgpu is still correctly interpreting this as a signed int, right?

p1-0tr replied:
I've not seen any evidence to the contrary. The standard says that literals without a suffix should be treated as `AbstractInt` (https://www.w3.org/TR/WGSL/#numeric-literal) with values in the range `[-2^63, 2^63)` (https://www.w3.org/TR/WGSL/#abstractint). So, I think the answer is yes :)

p1-0tr commented on crates/burn-wgpu/src/compute/webgpu_api.rs:
```diff
@@ -0,0 +1,190 @@
+    async fn select_adapter<G: GraphicsApi>(device: &WgpuDevice) -> Self::Adapter;
+    #[cfg(not(target_family = "wasm"))]
+    fn select_adapter<G: GraphicsApi>(device: &WgpuDevice) -> Self::Adapter;
+    fn read_buffer(buffer: &Self::Buffer, device: &Self::Device) -> Vec<u8>;
```
I ended up moving it to the buffer trait, since it does not use the queue.

nathanielsimard commented:
@p1-0tr Just an update:

As you may have seen, we updated quite a lot in the wgpu server, which solved a lot of issues with different graphics APIs and improved performance a bit. We also changed how the element types are handled in the JIT backend, which paves the way for quantization and our new GPU Rust API. We still want to go further with this PR and eventually merge it. Things are a bit more stable now, so we can better review and integrate it.

Note: I don't think using a git submodule is optimal; I would prefer downloading it with a tagged version of Dawn to the cache directory using [this method](https://github.com/tracel-ai/burn/blob/d3cd6c4928977ab7b345396b800fbbb75e34e72a/crates/burn-common/src/network.rs#L21), however this could be done in a following PR.

p1-0tr commented:
> As you may have seen, we updated quite a lot in the wgpu server, which solved a lot of issues with different graphics APIs and improved performance a bit. We also changed how the element types are handled in the JIT backend, which paves the way for quantization and our new GPU Rust API. We still want to go further with this PR and eventually merge it. Things are a bit more stable now, so we can better review and integrate it.

@nathanielsimard thanks for the update. I've noticed the changes when doing my last rebase :D

> Note: I don't think using a git submodule is optimal; I would prefer downloading it with a tagged version of Dawn to the cache directory using [this method](https://github.com/tracel-ai/burn/blob/d3cd6c4928977ab7b345396b800fbbb75e34e72a/crates/burn-common/src/network.rs#L21), however this could be done in a following PR.

Thanks :) I'll look into that.

laggui commented:
This belongs to cubecl now 😅 so I'll close due to the massive changes in this area.
